src/ast/passes/simplify_types.cpp

```diff
@@ -0,0 +1,130 @@
+#include <optional>
+
+#include "ast/ast.h"
+#include "ast/passes/simplify_types.h"
+#include "ast/visitor.h"
+#include "struct.h"
+#include "types.h"
+
+namespace bpftrace::ast {
+
+class SimplifyTypes : public Visitor<SimplifyTypes, std::optional<Expression>> {
+public:
+  SimplifyTypes(ASTContext &ast) : ast_(ast) {};
+
+  using Visitor<SimplifyTypes, std::optional<Expression>>::visit;
+
+  std::optional<Expression> visit(Binop &op);
+  std::optional<Expression> visit(Expression &expr);
+
+private:
+  Expression create_land_chain(std::vector<Expression> &equal_exprs, Binop &op);
+  ASTContext &ast_;
+};
+
+std::optional<Expression> SimplifyTypes::visit(Expression &expr)
+{
+  auto r = Visitor<SimplifyTypes, std::optional<Expression>>::visit(expr.value);
+  if (r) {
+    expr.value = r->value;
+  }
+  return std::nullopt;
+}
+
+Expression SimplifyTypes::create_land_chain(
+    std::vector<Expression> &equal_exprs,
+    Binop &op)
+{
+  auto end = equal_exprs.back();
+  equal_exprs.pop_back();
+
+  if (equal_exprs.empty()) {
+    return end;
+  }
+
+  auto *and_binop = ast_.make_node<Binop>(end,
+                                          Operator::LAND,
+                                          create_land_chain(equal_exprs, op),
+                                          Location(op.loc));
+  and_binop->result_type = CreateBool();
+
+  return and_binop;
+}
+
+std::optional<Expression> SimplifyTypes::visit(Binop &op)
+{
+  visit(op.left);
+  visit(op.right);
+
+  if (op.op != Operator::EQ && op.op != Operator::NE) {
+    return std::nullopt;
+  }
+
+  const SizedType &left_ty = op.left.type();
+  const SizedType &right_ty = op.right.type();
+
+  if (!left_ty.IsTupleTy() || !right_ty.IsTupleTy()) {
+    return std::nullopt;
+  }
+
+  auto &left_fields = left_ty.GetFields();
+  auto &right_fields = right_ty.GetFields();
+
+  if (left_fields.size() != right_fields.size()) {
+    // This is a semantic error
+    return std::nullopt;
+  }
+
+  if (left_fields.empty()) {
+    return ast_.make_node<Boolean>(op.op == Operator::EQ, Location(op.loc));
+  }
+
+  // N.B. Because of the way we're accessing each field of the tuple
+  // and essentially turning this:
+  // `$x = ("hello", -6); $y = ("bye", -6); $x == $y`
+  // into
+  // `$x = ("hello", -6); $y = ("bye", -6); ($x.0 == $y.0 && $x.1 == $y.1)`
+  // If the tuple contains an expression, it might not be evaluated
+  // (due to short-circuiting) if previous elements in the tuple are not equal.
+  // We're doing this instead of a generic memcmp because we can be less
+  // strict about the inner types of the tuple matching up due to size
+  // and alignment issus.
+  // https://github.com/bpftrace/bpftrace/pull/4523#discussion_r2382109017
+  std::vector<Expression> equal_exprs;
+  for (size_t i = 0; i < left_fields.size(); ++i) {
+    auto *left_tpa = ast_.make_node<TupleAccess>(op.left, i, Location(op.loc));
```

Comment:
The approach is good, but you do have a no-no here. This is copying the same expression value into multiple nodes, effectively "untree-ing" the AST, and potentially duplicated a complex expression.

What needs to happen instead, is that you need to either:

- Unpack the elem expr directly (don't wrap in an access) if this is a literal tuple
- clone the expression if this is a variable
- wrap the whole comparison in a block with a variable for anything else

These three apply to both the left and right, so you could end up with:

```
@[(x,y,z)] == (1,2,3)
=> {let $$tmp = @[(x,y,z)]; $$tmp.0 == 1 && $$tmp.1 == 2 && $$tmp.2 == 3}
```

(The block could have up to two variables, or none)

Reply:
@amscanne I think this might be the simplest solution (rather than inspecting if one side is a literal) - it also solves the issue of not evaluating all the elements due to possible short-circuiting:

- https://github.com/bpftrace/bpftrace/pull/4687